### ISSUE TITLE
Toil tmp directories are being uploaded in output

### ISSUE THREAD

**Issue by dleehr** [Duke-GCB/lando#68]
In a recent run, the toil tmpXXXXX directories are being included in the output upload. These should not be included.

**Comment by dleehr** [comment]
This is easily reproducible with cwltoil.

I created a simple packed workflow that just runs one tool (fastqc) and a simple JSON job order file with one fastq file.

```
cwltoil --logDebug --logFile debug.log --workDir /work/debug-dangling-tmpdirs/debug-workdir/ --jobStore /work/debug-dangling-tmpdirs/debug-jobStore/ --outdir /work/debug-dangling-tmpdirs/debug-results/ just-qc-packed.cwl#main onefile.json
```

toil cleans its own jobstore and the files it stages in and out, but it does leave `tmpXXXXX` directories beneath the specified outdir (`debug-results`):

```
./debug-workdir
./debug-dangling-tmpdir.json
./just-qc-packed.cwl
./debug.log
./debug.sh
./onefile.json
./debug-results
./debug-results/tmpbKBcSr
./debug-results/tmpbKBcSr/G-004-F_S27_L007_R2_001_fastqc.html
./debug-results/tmpbKBcSr/.cache
./debug-results/tmpbKBcSr/.cache/fontconfig
./debug-results/tmpbKBcSr/.cache/fontconfig/CACHEDIR.TAG
./debug-results/tmpbKBcSr/.cache/fontconfig/3830d5c3ddfd5cd38a049b759396e72e-le64.cache-6
./debug-results/tmpbKBcSr/.cache/fontconfig/d589a48862398ed80a3d6066f4f56f4c-le64.cache-6
./debug-results/tmpbKBcSr/.cache/fontconfig/4c599c202bc5c08e2d34565a40eac3b2-le64.cache-6
./debug-results/tmpbKBcSr/.cache/fontconfig/7ef2298fde41cc6eeb7af42e48b7d293-le64.cache-6
./debug-results/tmpbKBcSr/G-004-F_S27_L007_R2_001_fastqc.zip
./debug-results/G-004-F_S27_L007_R2_001_fastqc.zip

Looking at the debug.log file, it seems that this is the tmp directory that CWL creates to bind as `/var/spool/cwl` inside the docker container.

```
[job fastqc.cwl] /work/debug-dangling-tmpdirs/debug-results/tmpbKBcSr$ docker \
    run \
    -i \
    --volume=/work/debug-dangling-tmpdirs/debug-workdir/toil-f8bcc45c-58ef-4276-8787-523e28b4e9df-0a1a20a2548749299d131a4d6c575d3b/tmpeHtNdY/3f48c399-21b0-4b2d-af83-5c5d797745e1/tmp2DMAhZ.tmp:/var/lib/cwl/stgca9e14b4-1721-418a-a3a0-e0a39216ffdc/G-004-F_S27_L007_R2_001.fastq.gz:ro \
    --volume=/work/debug-dangling-tmpdirs/debug-results/tmpbKBcSr:/var/spool/cwl:rw \
    --volume=/work/debug-dangling-tmpdirs/debug-workdir/toil-f8bcc45c-58ef-4276-8787-523e28b4e9df-0a1a20a2548749299d131a4d6c575d3b/tmpeHtNdY/3f48c399-21b0-4b2d-af83-5c5d797745e1/tTCG1dz/tmp:/tmp:rw \
    --workdir=/var/spool/cwl \
    --read-only=true \
    --rm \
    --env=TMPDIR=/tmp \
    --env=HOME=/var/spool/cwl \
    dukegcb/fastqc:0.11.4 \
    fastqc \
    --noextract \
    --format \
    fastq \
    /var/lib/cwl/stgca9e14b4-1721-418a-a3a0-e0a39216ffdc/G-004-F_S27_L007_R2_001.fastq.gz \
    --dir \
    /tmp \
    -o \
    /var/spool/cwl \
    --threads \
    8
```

So the docker container is placing its files there as expected, and the desired CWL output file is getting pulled out correctly, but the rest of the junk leftover in the directory is still there.

**Comment by dleehr** [comment]
Next I'll see if this happens with cwltool or if it happens when we don't specify the `--outdir`

**Comment by dleehr** [comment]
Before even completing I can tell this won't be an issue for cwltool. It puts temporary output directories in `/tmp`:

```
[job qc] /tmp/tmpwb3kwP$ docker \
    run \
    -i \
    --volume=/work/data_for_job_41/G-004-F_S27_L007_R2_001.fastq.gz:/var/lib/cwl/stgf7fefe7e-ca30-4e8d-992b-1128a753f3ca/G-004-F_S27_L007_R2_001.fastq.gz:ro \
    --volume=/tmp/tmpwb3kwP:/var/spool/cwl:rw \
    --volume=/tmp/tmpjigEe0:/tmp:rw \
    --workdir=/var/spool/cwl \
    --read-only=true \
    --user=0 \
    --rm \
    --env=TMPDIR=/tmp \
    --env=HOME=/var/spool/cwl \
    dukegcb/fastqc:0.11.4 \
    fastqc \
    --noextract \
    --format \
    fastq \
    /var/lib/cwl/stgf7fefe7e-ca30-4e8d-992b-1128a753f3ca/G-004-F_S27_L007_R2_001.fastq.gz \
    --dir \
    /tmp \
    -o \
    /var/spool/cwl \
    --threads \
    8
```

**Comment by dleehr** [comment]
https://github.com/BD2KGenomics/toil/blob/4de2743762bc3ce810c6fce195547aad85570242/src/toil/cwl/cwltoil.py#L431

Looks like this is happening because we're specifying `--outdir` 😠

**Comment by dleehr** [comment]
actually that commit was only a few days ago, so it wouldn't be impacting yet. But there has been some movement around the tmp outdir

**Comment by dleehr** [comment]
In the version I'm using (3.9.1) and the current master, cwltoil sets `tmpdir` to `outdir`

https://github.com/BD2KGenomics/toil/blob/cda81ab77d3681276b435932b044fec9fa4d9cde/src/toil/cwl/cwltoil.py#L746

Will file an issue on Toil though

**Comment by dleehr** [comment]
Looks like this is fixed in the current development version of toil. The docker container uses a directory from the `workdir` as `/var/spool/cwl`:

```
"/work/debug-dangling-tmpdirs/debug-workdir/out_tmpdirWakMV0:/var/spool/cwl:rw",
```

I'll test to see if fixed in 3.10 or if we need to wait for a new release.

**Comment by dleehr** [comment]
This is indeed fixed in 3.10, which is the current version on pypi. I'll update our lando worker playbook to specify this version.
